### ISSUE TITLE
"Headless" ICF - API access to functionality

### ISSUE THREAD

**Issue by ahpook** [github-community-projects/internal-contribution-forks#120]
### Is your feature request related to a problem?

Some users are working to integrate ICF into a larger "open source portal" context, where they already have a web app and additional tooling around OSS contributions (like which projects are available for contributions, who is allowed to contribute, etc). They'd like to make use of the internals of ICF - the mirror lifecycle management and synchronization - but wrap this in their own web service.

### Describe the solution you'd like

The goal would be to make ICF "embeddable" in other applications. @ajhenry brainstormed a couple of options like creating a RPC client, making the app "import"-able as a package, or having a separate web service from the front end which could be called by another app. Not sure which has the highest effort/reward and would best meet the user need, so this requires more exploration and thought.

**Comment by Miablo** [comment]
> Some users are working to integrate ICF into a larger "open source portal" context, where they already have a web app and additional tooling around OSS contributions (like which projects are available for contributions, who is allowed to contribute, etc).

For Private Mirrors to have a 'headless' API functionality some of the workflows would change when the currently coupled UI is replaced with the user's larger 'open source portal', right? 

Would this issue describe the creation of a stand-alone API that can then be onboarded internally by a team who does have that internal facing portal? The assumptions made here are that the user would:

1. have a db or some system of record of who is allowed to contribute
2. have a db or system of record of what projects can be contributed to
3. have a list of licenses or other parameters that will need to be checked prior to allow a contribution be made to a repo
4. require new user registration would be handled by the existing 'open source portal'
5. have SSO integration would be used instead of the OAuth w/ GitHub method used by the current application to verify the user's identify and then verify their access to the various orgs

Currently there are four routers: 
1. Config
2. Git
3. Repos
4. Octokit

I would assume that these would become the end points or callable routes in this API? Would some additional functionality be useful for audit and/or compliance checks? (--Just starting to think out loud here--)

**Comment by ahpook** [comment]
@Miablo thanks for the write-up!

> For Private Mirrors to have a 'headless' API functionality some of the workflows would change when the currently coupled UI is replaced with the user's larger 'open source portal', right?

Yes I think that's exactly right, and as you say at the bottom the different routers would become callable separately, just to execute the functionality they are responsible for, and the larger user auth/authz would be externalized since it's likely different for everyone.

**Comment by Miablo** [comment]
@ahpook What do you envision for this implementation? Are you ya'll locked in on using tRPC or would you all be open to using something like:

1. [DeepKit](https://deepkit.io/)
2. [NestJs](https://nestjs.com/) with[fastify](https://fastify.dev/)
3. or anything else that may be a good fit

I'm trying to think through the scope for this specific issue and it would make sense to have one issue (this current one) to create the API itself and then other new issues, if ya'll want, to create a solo component for the UI and a solo component for the web hooks. This would then create separate consumable components of private mirrors or provide the full functioning app as well as an option for consumers to pick and choose their solution. 

What are your thoughts?

**Comment by Miablo** [comment]
Thank you, thank you for the keys to the kingdom! 😁  I will create a proposed API design doc / diagram to toss in here for review.

**Comment by Miablo** [comment]
@ahpook -- we just had a design session internally and talked about how for our initial pilot to use PMA we, OSPO, will be manually creating the forks for our testers. Because of this  I wanted to propose adding a `/forks` endpoint to this new PMA API. The idea is that we can `GET`, `POST`, `PATCH/PUT`, or `DELETE` forks using the endpoint. More specifically though I wanted to focus on the `POST` method where we would create the fork.

I created a draft of a diagram to show how this endpoint could run through a few validations before creating the fork. The initial assumption is that if this `create a fork` button or whatever is used to trigger this endpoint is `enabled` then the user has successfully passed authentication checks (logged in to their gh account, etc.).

I wanted to get the conversation started regarding this endpoint. The area that is encapsulated in the diagram houses additional checks we could do as part of the fork creation request. 

One big one is the license check and to block or prevent contributions from being made to projects who don't have approved licenses. 

I would like to make these checks module somehow where a company can input a list or some other data source to use for approved licenses or approved projects they may have, etc. 

![image](https://github.com/user-attachments/assets/d794d2eb-709e-4bc1-92d4-1a4281677a5e)

Thoughts?

**Comment by Miablo** [comment]
> Do you expect the policy checks in the Fork exists -> No branch to be configurable or extensible? I.e. how would someone else with a different set of policies implement them?

Yes! That is how I imagined it in my mind, but wasn't sure how to implement. I wonder if I can create some example or common custom checks that could be used by others as a template using git-proxy? I need to look into it more and see how git-proxy would work, but looks promising!

**Comment by Miablo** [comment]
https://github.com/finos/git-proxy/issues/594 <--- Looks like someone from Citi is working on adding the license check plugin 🥇

**Comment by wrslatz** [comment]
Been thinking on this a bit over the last few weeks... apart from a fully separate API or importable package, I wonder if a simpler initial approach could be to create [Next.js API Routes](https://nextjs.org/docs/pages/building-your-application/routing/api-routes) to expose app functionality via API. This would align most closely with the current app functionality, keeping the project simpler and likely reusing more code. It would enable implementing #121, which is something I am interested in revisiting in the near future, as CLI access is a standard pattern for agentic software development.

The UI and API portions of the app could each be gated behind configuration, allowing end users to enable only the parts of the app that they specifically wish to use. For end users that wish to wrap PMA behind their own web service, this could potentially be done in other ways: using a proxy service or hosting a separate service in front of PMA.

I still think exposing a library for PMA functionality could be useful, but would be a larger undertaking. Exposing an API as part of PMA could be a good initial step towards that final solution.

**Comment by riley-kohler** [comment]
This came up in some discussions between @Miablo and I as well. I think focusing on #121 and as part of that genericising some of the logic to be used by the CLI makes sense. I'm not too familiar with Next API Routes to know if working on that first to enable the CLI or working on the CLI to peel out individual logic into a shared package makes the most sense.

**Comment by riley-kohler** [comment]
Authentication is going to be one of the biggest problems we'll need to deal with and I'm not sure the best way to do that with an accessible API. I guess we just let users pass in a token. Focusing on the cli extension could alleviate some of that by making use of the built in gh auth.

**Comment by wrslatz** [comment]
> This came up in some discussions between [@Miablo](https://github.com/Miablo) and I as well. I think focusing on [#121](https://github.com/github-community-projects/private-mirrors/issues/121) and as part of that genericising some of the logic to be used by the CLI makes sense. I'm not too familiar with Next API Routes to know if working on that first to enable the CLI or working on the CLI to peel out individual logic into a shared package makes the most sense.

Do we think a CLI would interact directly with GitHub and perform the actions that the app currently takes via the UI or would the CLI call Private Mirrors API endpoints? I am thinking the latter would be necessary, as the App is likely to have permissions and would be necessary to implement checks and limit direct user permissions.

> Authentication is going to be one of the biggest problems we'll need to deal with and I'm not sure the best way to do that with an accessible API. I guess we just let users pass in a token. Focusing on the cli extension could alleviate some of that by making use of the built in gh auth.

If the solution we end up going with is "expose Private Mirrors functionality as API routes" then the authentication solution would likely be the same as with the UI. Users would authenticate via the GitHub App using OAuth. Looking for examples of gh CLI extensions that follow this pattern could be useful.

**Comment by wrslatz** [comment]
As an aside, as part of #411, I noticed that the tests in [`test/app.ts`](https://github.com/github-community-projects/private-mirrors/blob/main/test/app.ts) and [`test/app.test.ts`](https://github.com/github-community-projects/private-mirrors/blob/main/test/app.test.ts) both felt somewhat overlapping and somewhat larger than unit tests. As part of implementing this issue, or maybe alongside #428, I'd like to get unit tests separated from UI and API tests that test the app end to end more fully. I'll extract this out to its own issue.

**Comment by wrslatz** [comment]
Maybe the next best steps for this issue and #121 would be
1. identify, at a high level, each action that we expect a user to be able to invoke via API/gh CLI and the corresponding app functionality that exists today
2. consider the permissions needed for each action
3. determine whether the user should be able to take the action themselves or whether the app needs to take the action on their behalf

Thoughts?

**Comment by riley-kohler** [comment]
> Do we think a CLI would interact directly with GitHub and perform the actions that the app currently takes via the UI or would the CLI call Private Mirrors API endpoints? I am thinking the latter would be necessary, as the App is likely to have permissions and would be necessary to implement checks and limit direct user permissions.
 
Thinking out loud: The main functionality that the UI performs that we would want to expose through a CLI is mirror creation. I think that should be simple enough for the CLI extension to perform directly using the same function as the current API but as a different entrypoint. I could see pointing the CLI at the desired hosted API being tricky since there is not one central API that all customers will use as a default. The app permissions are necessary for syncing that is currently performed automatically though we want to expose a manual sync in the future so maybe it is necessary to go through the API for that reason. Even then both routes could make sense for each bit of functionality. 

> If the solution we end up going with is "expose Private Mirrors functionality as API routes" then the authentication solution would likely be the same as with the UI. Users would authenticate via the GitHub App using OAuth. Looking for examples of gh CLI extensions that follow this pattern could be useful.

Yeah, I think this is my biggest uncertainty. The sense that I got is most gh cli extensions just use user authentication to hit the GitHub API but examples of a CLI that is mainly a shell to hit a bespoke API would be helpful. Currently I think the permission structure of the Private Mirrors needs be refined. We are basically letting users gain a specific set of privileges from the private mirrors app by authenticating as their user so we need to be careful what we expose to avoid effectively granting the user those privileges. An example is that users should likely only be able to create mirrors for forks that their user has access to rather than what private mirrors has access to since private forks could potentially be sensitive.

**Comment by wrslatz** [comment]
I agree with you that having a user directly perform actions gets tricky from an access and permissions perspective. Exposing an API from PMA would keep everything more aligned to how the UI works today, simplifying docs and setup.

A gh CLI extension could just take in as config the URL for the hosted PMA App, OAuth on behalf of the user like the current UI does, and then call the hosted PMA App APIs using the OAuth credentials.

Seems like the next step is to investigate if there are any gh CLI extensions that act upon a separately hosted GitHub App, since that is effectively what would be implemented here.

**Comment by riley-kohler** [comment]
> Maybe the next best steps for this issue and [#121](https://github.com/github-community-projects/private-mirrors/issues/121) would be
> 
> 1. identify, at a high level, each action that we expect a user to be able to invoke via API/gh CLI and the corresponding app functionality that exists today
> 2. consider the permissions needed for each action
> 3. determine whether the user should be able to take the action themselves or whether the app needs to take the action on their behalf
> 
> Thoughts?

> Seems like the next step is to investigate if there are any gh CLI extensions that act upon a separately hosted GitHub App, since that is effectively what would be implemented here.

I'm aligned with these next steps. If we have time we can discuss further as part of the community sync today.

**Comment by Miablo** [comment]
> > Maybe the next best steps for this issue and [#121](https://github.com/github-community-projects/private-mirrors/issues/121) would be
> > 
> > 1. identify, at a high level, each action that we expect a user to be able to invoke via API/gh CLI and the corresponding app functionality that exists today
> > 2. consider the permissions needed for each action
> > 3. determine whether the user should be able to take the action themselves or whether the app needs to take the action on their behalf
> > 
> > Thoughts?
> 
> > Seems like the next step is to investigate if there are any gh CLI extensions that act upon a separately hosted GitHub App, since that is effectively what would be implemented here.
> 
> I'm aligned with these next steps. If we have time we can discuss further as part of the community sync today.

Did y'all get to this today? I missed the sync 😭

**Comment by Miablo** [comment]
Also for this part: 
> identify, at a high level, each action that we expect a user to be able to invoke via API/gh CLI and the corresponding app functionality that exists today

Questions:
The user we think will be using the CLI will be a person looking to make a contribution? 
For the API would this be a group (like an OSPO) looking to use PMA functionality without the UI?
Or some combo?

For Actions this is the initial list I can think of with a proposed table structure to outline who should do what

| Action | Actor (who/what should do this) | Permission Required |
| -- | -- | -- |
| Get Org Configs |  | |
| Check Org Membership | | |
| Add to Org | | |
| Remove from Org | | |
| List Forks | | |
| Create Fork | | |
| Edit Fork | || 
| Delete Fork | | |
| Create Mirror | | |
| Edit Mirror | | |
| Delete Mirror | | |
| List Mirrors | | |
| Sync Fork | | |
| Sync Mirror | | |

**Comment by wrslatz** [comment]
> Did y'all get to this today? I missed the sync 😭 

We did not, so no worries!

> Questions:
> The user we think will be using the CLI will be a person looking to make a contribution? 
> For the API would this be a group (like an OSPO) looking to use PMA functionality without the UI?
> Or some combo?

I was primarily thinking from a contributing user perspective. The API could be useful for OSPOs or people hosting and managing a PMA instance, allowing easier extension and building on top of PMA, or otherwise using it where the UI experience is not sufficient or desired.

Assuming we get to API and CLI, my next idea was going to be creating a skill to instruct AI agents on how to use the PMA CLI. I was going to use the Playwright CLI and accompanying skill as inspiration.
